{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import six.moves.urllib.request as urlreq\n",
    "from six import PY3\n",
    "import json\n",
    "import dash\n",
    "import dash_bio as dashbio\n",
    "import dash_html_components as html\n",
    "import json\n",
    "import six.moves.urllib.request as urlreq\n",
    "from six import PY3\n",
    "from dash_bio_utils import circos_parser as cp\n",
    "\n",
    "import dash\n",
    "import dash_bio as dashbio\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "\n",
    "external_stylesheets = ['https://codepen.io/chriddyp/pen/bWLwgP.css']\n",
    "\n",
    "app = dash.Dash(__name__, external_stylesheets=external_stylesheets)\n",
    "\n",
    "data = urlreq.urlopen(\n",
    "    'https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/' +\n",
    "    'circos_graph_data.json'\n",
    ").read()\n",
    "\n",
    "model_data = urlreq.urlopen(\n",
    "    'https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/' +\n",
    "    'mol3d/model_data.js'\n",
    ").read()\n",
    "\n",
    "\n",
    "styles_data = urlreq.urlopen(\n",
    "    'https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/' +\n",
    "    'mol3d/styles_data.js'\n",
    ").read()\n",
    "\n",
    "\n",
    "if PY3:\n",
    "    model_data = model_data.decode('utf-8')\n",
    "    styles_data = styles_data.decode('utf-8')\n",
    "    data = data.decode('utf-8')\n",
    "\n",
    "model_data = json.loads(model_data)\n",
    "styles_data = json.loads(styles_data)\n",
    "circos_graph_data = json.loads(data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "#make circos features\n",
    "\n",
    "#chrom\tchromStart\tchromEnd\tname\tgieStain\n",
    "#chr1\t12700000\t16200000\tp36.21\tgpos50\n",
    "\n",
    "vals = 'gneg gpos50 gpos75 gpos25 gpos100 acen gvar stalk'.split()\n",
    "\n",
    "annotationfile = './coronaGenoWuhan.gb'\n",
    "parser = SeqIO.parse( annotationfile , 'gb')\n",
    "\n",
    "\n",
    "#color the orfs \n",
    "\n",
    "#\n",
    "\n",
    "#add the blast results to the circos \n",
    "with open('corona_circos.txt' , 'w') as circout:\n",
    "    circout.write('#chrom\tchromStart\tchromEnd\tname\tgieStain\tcolor')\n",
    "    for entry in parser:\n",
    "        for feature in entry.features:\n",
    "            if 'product' in feature.qualifiers:\n",
    "                line = 'chr1\\t' + str(feature.location.start)+'\\t' + str(feature.location.end) +'\\t' + feature.qualifiers['product'][0] + '\\t' + vals[1] + '\\t #CC0000'\n",
    "                circout.write(line )\n",
    "                \n",
    "json_circos = cp.txt_to_track('corona_circos.txt')\n",
    "print(json_circos)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38\n",
      "['#ff0000', '#ff0700', '#ff0d00', '#ff1400', '#ff1b00', '#ff2200', '#ff2800', '#ff2f00', '#ff3600', '#ff3c00', '#ff4300', '#ff4a00', '#ff5100', '#ff5700', '#ff5e00', '#ff6500', '#ff6b00', '#ff7200', '#ff7900', '#ff7f00', '#ff8600', '#ff8d00', '#ff9400', '#ff9a00', '#ffa100', '#ffa800', '#ffae00', '#ffb500', '#ffbc00', '#ffc300', '#ffc900', '#ffd000', '#ffd700', '#ffdd00', '#ffe400', '#ffeb00', '#fff200', '#fff800', '#ffff00']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pandas as pd\n",
    "from itertools import combinations\n",
    "import numpy as np\n",
    "from colour import Color\n",
    "count = 0\n",
    "\n",
    "annotationfile = './coronaGenoWuhan.gb'\n",
    "parser = SeqIO.parse( annotationfile , 'gb')\n",
    "\n",
    "\n",
    "### stack graph\n",
    "for entry in parser:\n",
    "    for feature in entry.features:\n",
    "        if 'product' in feature.qualifiers:\n",
    "            count +=1\n",
    "\n",
    "#{'chr': 'chr9', 'start': 2728594, 'end': 2948764, 'block_id': 'chr9'}\n",
    "color1 = Color('red')\n",
    "color2 = Color('yellow')\n",
    "print(count)\n",
    "colorvec = color1.range_to(color2, count+1)\n",
    "colorvec = [ c.hex_l for c in colorvec ]\n",
    "print(colorvec)\n",
    "###layout fmt\n",
    "#{'id': 'chr1', 'label': 'chr1', 'color': '#996600', 'len': 249250621 }\n",
    "circos_graph_data = {}\n",
    "circos_graph_data['layout'] = [{'id': 'chr1', 'label':'' , 'len': 29674 , 'color':'#CC0000' } ]\n",
    "circos_graph_data['stack'] = []\n",
    "parser = SeqIO.parse( annotationfile , 'gb')\n",
    "\n",
    "count = 0\n",
    "featlist = []\n",
    "for entry in parser:\n",
    "    for feature in entry.features:\n",
    "        if 'product' in feature.qualifiers:\n",
    "            count +=1\n",
    "            if feature.qualifiers['product'][0] not in featlist:\n",
    "                featlist.append(feature.qualifiers['product'][0])\n",
    "                circos_graph_data['stack'].append( {'chr': feature.qualifiers['product'][0], 'block_id':'chr1' , 'start': feature.location.start , 'end':  feature.location.end , 'color':colorvec[count] } )    \n",
    "clusterdf = pd.read_csv( '../lucy_mk2/30_07_2020/gisaid_hcov-2020_07_30.QC.NSoutlier.filter.deMaiomask.EPIID.HF.noambig.aln_svd_codon_clustersclusterpositions_columns.csv')\n",
    "\n",
    "#one color per cluster\n",
    "for i,r in clusterdf.iterrows():    \n",
    "    rowpos = [ int(entry) for entry in r if not np.isnan(entry) ]\n",
    "    circos_graph_data[i] = [ {'color': '#ff5722', 'source': {'id': 'chr1', 'start': i, 'end': i+2 }, 'target': {'id': 'chr1', 'start': j, 'end': j+2}} for i,j in combinations(rowpos,2) ] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "models = '5x58,6cs0,6sc1,6nb6, 2jw8, 2xab, 4aud1, 1yo4, 2acf, 2wct, 3vc8, 2gt7, 3ee7, 2g9t, 3ee7, 2g9t, 6jyt, 1ysy, 6nur, 2g9t, 5c8u, 2g9t, 2xyq, 4mm3, 6cs2, 6acg, 6acj, 6ack, 2dd8, 2ghw, 6nb6, 6nb7'\n",
    "models = models.split(',')\n",
    "templatedir = './templates/'\n",
    "\n",
    "#map structs to geno. \n",
    "\n",
    "#add track on geno\n",
    "\n",
    "#dict from geno 2 struct residue\n",
    "\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import wget\n",
    "import requests\n",
    "import glob\n",
    "import time\n",
    "clear = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isdir(templatedir):\n",
    "    os.mkdir(templatedir)\n",
    "if clear == True:\n",
    "    files = glob.glob(templatedir+'*.pdb')\n",
    "    for f in files:\n",
    "        os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['./templates/6ACG.pdb', './templates/6NUR.pdb', './templates/5C8U.pdb', './templates/2DD8.pdb', './templates/2WCT.pdb', './templates/6ACJ.pdb', './templates/2G9T (1).pdb', './templates/6NB6.pdb', './templates/5X58.pdb', './templates/6NB7.pdb', './templates/6JYT.pdb', './templates/2GT7.pdb', './templates/3VC8.pdb', './templates/3EE7.pdb', './templates/4MM3.pdb', './templates/2ACF.pdb', './templates/2G9T.pdb', './templates/2XYQ.pdb', './templates/6NB6 (1).pdb', './templates/6CS2.pdb', './templates/6CS0.pdb', './templates/1YSY.pdb', './templates/1YO4.pdb', './templates/2XAB.pdb', './templates/6ACK.pdb', './templates/2GHW.pdb', './templates/2JW8.pdb', './templates/3EE7 (1).pdb']\n",
      "6sc1\n",
      "err 6sc1\n",
      " 4aud1\n",
      "err  4aud1\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import *\n",
    "dl_url = 'http://files.rcsb.org/download/'\n",
    "dl_url_err = 'http://files.rcsb.org/download/'\n",
    "\n",
    "structs = {}\n",
    "already = glob.glob( './templates/*.pdb' )\n",
    "print(already)\n",
    "#pull complexes\n",
    "for m in models:\n",
    "    structfile = './templates/'+m.upper().strip()+'.pdb'\n",
    "    if structfile not in already:\n",
    "        print(m)\n",
    "        time.sleep(1)\n",
    "        try:\n",
    "            wget.download(url = dl_url + m.strip() +'.pdb' , out =structfile)\n",
    "            structs[m] = structfile\n",
    "        except:\n",
    "            try:\n",
    "                wget.download(url = dl_url + m.strip() +'.pdb' , out =structfile)\n",
    "                structs[m] = structfile\n",
    "            except:\n",
    "                print('err', m )\n",
    "    else:\n",
    "        structs[m.strip()] = structfile\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'5x58': './templates/5X58.pdb', '6cs0': './templates/6CS0.pdb', '6nb6': './templates/6NB6.pdb', '2jw8': './templates/2JW8.pdb', '2xab': './templates/2XAB.pdb', '1yo4': './templates/1YO4.pdb', '2acf': './templates/2ACF.pdb', '2wct': './templates/2WCT.pdb', '3vc8': './templates/3VC8.pdb', '2gt7': './templates/2GT7.pdb', '3ee7': './templates/3EE7.pdb', '2g9t': './templates/2G9T.pdb', '6jyt': './templates/6JYT.pdb', '1ysy': './templates/1YSY.pdb', '6nur': './templates/6NUR.pdb', '5c8u': './templates/5C8U.pdb', '2xyq': './templates/2XYQ.pdb', '4mm3': './templates/4MM3.pdb', '6cs2': './templates/6CS2.pdb', '6acg': './templates/6ACG.pdb', '6acj': './templates/6ACJ.pdb', '6ack': './templates/6ACK.pdb', '2dd8': './templates/2DD8.pdb', '2ghw': './templates/2GHW.pdb', '6nb7': './templates/6NB7.pdb'}\n"
     ]
    }
   ],
   "source": [
    "print(structs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ALA': 'A', 'ASX': 'B', 'CYS': 'C', 'ASP': 'D', 'GLU': 'E', 'PHE': 'F', 'GLY': 'G', 'HIS': 'H', 'ILE': 'I', 'LYS': 'K', 'LEU': 'L', 'MET': 'M', 'ASN': 'N', 'PRO': 'P', 'GLN': 'Q', 'ARG': 'R', 'SER': 'S', 'THR': 'T', 'SEC': 'U', 'VAL': 'V', 'TRP': 'W', 'XAA': 'X', 'TYR': 'Y', 'GLX': 'Z'}\n"
     ]
    }
   ],
   "source": [
    "#open all the complexes and pull sequences\n",
    "#convert to json\n",
    "from Bio import Alphabet\n",
    "parser = PDBParser()\n",
    "from Bio.SeqUtils import seq1\n",
    "letter3 = Alphabet.ThreeLetterProtein.letters\n",
    "converter = { l.upper(): seq1(l) for l in letter3 }\n",
    "print(converter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 26118.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 26314.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 26510.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 26692.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 26960.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 27242.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 27690.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 28327.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 28948.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 3999.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 4021.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 4043.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 4289.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 6114.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 6126.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 6144.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 6180.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 6186.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 6414.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 6674.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 6906.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 8438.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 8441.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 8442.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 1708.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 1776.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 5299.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 5591.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 4120.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 4165.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 4182.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 4211.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 4256.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 4283.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 4308.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 4343.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 23721.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 23723.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 23725.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 23727.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain E is discontinuous at line 23729.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain F is discontinuous at line 23731.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain G is discontinuous at line 23733.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain H is discontinuous at line 23735.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain I is discontinuous at line 23737.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain J is discontinuous at line 23739.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain K is discontinuous at line 23741.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain L is discontinuous at line 23743.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain M is discontinuous at line 23745.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain N is discontinuous at line 23747.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain O is discontinuous at line 23749.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain P is discontinuous at line 23751.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain Q is discontinuous at line 23753.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain R is discontinuous at line 23755.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain S is discontinuous at line 23757.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain T is discontinuous at line 23759.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain U is discontinuous at line 23761.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain V is discontinuous at line 23763.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain W is discontinuous at line 23765.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain X is discontinuous at line 23767.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 23769.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 23838.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 23897.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 23963.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain E is discontinuous at line 24026.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain F is discontinuous at line 24089.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain G is discontinuous at line 24147.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain H is discontinuous at line 24212.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain I is discontinuous at line 24271.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain J is discontinuous at line 24326.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain K is discontinuous at line 24399.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain L is discontinuous at line 24470.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain M is discontinuous at line 24541.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain N is discontinuous at line 24615.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain O is discontinuous at line 24666.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain P is discontinuous at line 24711.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain Q is discontinuous at line 24769.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain R is discontinuous at line 24850.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain S is discontinuous at line 24911.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain T is discontinuous at line 24978.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain U is discontinuous at line 25040.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain V is discontinuous at line 25117.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain W is discontinuous at line 25166.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain X is discontinuous at line 25211.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 10163.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 10166.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 10169.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 10238.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 9512.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 21258.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 21262.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 21270.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 21274.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 3843.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 3874.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 3876.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 4100.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 3529.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 3560.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 28874.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 29275.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 29665.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain H is discontinuous at line 5314.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain S is discontinuous at line 5319.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain H is discontinuous at line 5333.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain L is discontinuous at line 5457.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain S is discontinuous at line 5567.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 7486.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 7487.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 7599.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 7717.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain D is discontinuous at line 7832.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain A is discontinuous at line 27655.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain B is discontinuous at line 28306.\n",
      "\n",
      "/home/cactuskid13/miniconda3/envs/pyprofiler3/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning:\n",
      "\n",
      "WARNING: Chain C is discontinuous at line 28880.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./structs.fast\n",
      " \"10 qseqid sseqid qlen slen qstart qend qframe evalue  \" \n",
      "structDB\n",
      "          qseqid    sseqid   qlen  slen  qstart   qend  qframe        evalue  \\\n",
      "0    NC_045512.2  6nur|A|0  29903   779   13789  16125       1  0.000000e+00   \n",
      "1    NC_045512.2  5c8u|D|0  29903   453   18040  19398       1  0.000000e+00   \n",
      "2    NC_045512.2  5c8u|B|0  29903   453   18040  19398       1  0.000000e+00   \n",
      "3    NC_045512.2  5x58|C|1  29903   419   22334  23590       2  0.000000e+00   \n",
      "4    NC_045512.2  5x58|B|1  29903   419   22334  23590       2  0.000000e+00   \n",
      "..           ...       ...    ...   ...     ...    ...     ...           ...   \n",
      "280  NC_045512.2  6jyt|B|2  29903    40   16936  17055       1  4.740000e-22   \n",
      "281  NC_045512.2  6jyt|A|2  29903    40   16936  17055       1  4.740000e-22   \n",
      "282  NC_045512.2  2g9t|V|1  29903    37   13169  13279       2  2.250000e-21   \n",
      "283  NC_045512.2  2g9t|F|1  29903    37   13169  13279       2  2.250000e-21   \n",
      "284  NC_045512.2  2g9t|U|1  29903    36   13172  13279       2  9.040000e-21   \n",
      "\n",
      "    struct chain  \n",
      "0     6nur     A  \n",
      "1     5c8u     D  \n",
      "2     5c8u     B  \n",
      "3     5x58     C  \n",
      "4     5x58     B  \n",
      "..     ...   ...  \n",
      "280   6jyt     B  \n",
      "281   6jyt     A  \n",
      "282   2g9t     V  \n",
      "283   2g9t     F  \n",
      "284   2g9t     U  \n",
      "\n",
      "[217 rows x 10 columns]\n",
      "217  orfs detected\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import shlex\n",
    "import pandas as pd \n",
    "#collapse codons\n",
    "#use tblastn to map to prot\n",
    "\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from Bio.PDB.Polypeptide import PPBuilder\n",
    "ppb=PPBuilder()\n",
    "\n",
    "with open( 'structs.fast' , 'w')as fastout:\n",
    "    for s in structs:\n",
    "        Structure = PDBParser().get_structure(s, structs[s])\n",
    "        for model in Structure:\n",
    "            for chain in model:\n",
    "                for i, pp in enumerate(ppb.build_peptides(chain)):\n",
    "                    fastout.write('>' + s + '|'+ chain.id +'|'+str(i)+'\\n')\n",
    "                    fastout.write(str( pp.get_sequence() ) +'\\n'  )\n",
    "\n",
    "\n",
    "def makeblastDB(fasta , blastdb):\n",
    "    print(fasta)\n",
    "    #make a blast database with the structures\n",
    "    args = 'makeblastdb -in '+ fasta +' -out '+ blastdb + ' -dbtype prot'\n",
    "    p = subprocess.run( shlex.split(args) )\n",
    "    return p ,  blastdb\n",
    "\n",
    "\n",
    "def runblastx( qfile ,db ,  outannot = 'outannot.txt' , outfmt = None ):\n",
    "    if outfmt is None:\n",
    "        outfmt = [ 'qseqid' , 'sseqid' , 'qlen' ,  'slen' , 'qstart' , 'qend' ,  'qframe' , 'evalue' ]\n",
    "        outfmt =  ' \"10 ' + ''.join([fmt+ ' ' for fmt in outfmt]) + ' \" '\n",
    "        print(outfmt)\n",
    "        print(db)\n",
    "    args = 'blastx -query '+ qfile + ' -db '+ db +' -outfmt' + outfmt + ' -out ' + outannot  \n",
    "    p = subprocess.run( shlex.split(args) )\n",
    "    return p , outannot\n",
    "\n",
    "\n",
    "out = 'outannot.txt'\n",
    "p,blastdb = makeblastDB( './structs.fast' , 'structDB' )\n",
    "p,annot = runblastx( 'covid19geno.fasta' , blastdb )\n",
    "\n",
    "annotation = pd.read_csv( annot , header = None )\n",
    "annotation.columns = [ 'qseqid' , 'sseqid' , 'qlen' ,  'slen' , 'qstart' , 'qend' ,  'qframe' , 'evalue' ] \n",
    "annotation = annotation[ annotation['evalue'] < 10**-3 ]\n",
    "annotation = annotation[annotation.evalue < 10 ** -20 ]\n",
    "annotation = annotation.drop_duplicates(['sseqid'])\n",
    "annotation['struct'] = annotation.sseqid.map(lambda x : x.split('|')[0])\n",
    "annotation['chain'] = annotation.sseqid.map(lambda x : x.split('|')[1])\n",
    "print(annotation)\n",
    "\n",
    "print(len(annotation), ' orfs detected')\n",
    "#recluster codons\n",
    "#load blast results into df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "circos_graph_data['struct_stack'] = []\n",
    "for i,r in annotation.iterrows():\n",
    "    circos_graph_data['struct_stack'].append( {'chr': r.sseqid , 'block_id':'chr1' , 'start': r.qstart , 'end':  r.qend  } ) \n",
    "    count +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import base64\n",
    "import json\n",
    "import tempfile\n",
    "\n",
    "from shutil import copy2\n",
    "from textwrap import dedent as s\n",
    "\n",
    "from dash.dependencies import Input, Output, State\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "import dash_daq as daq\n",
    "from dash_bio_utils import pdb_parser as parser, styles_parser as sparser\n",
    "import dash_bio\n",
    "\n",
    "\n",
    "def files_data_style(content):\n",
    "    fdat = tempfile.NamedTemporaryFile(suffix=\".js\", delete=False, mode='w+')\n",
    "    fdat.write(content)\n",
    "    dataFile = fdat.name\n",
    "    fdat.close()\n",
    "    return dataFile\n",
    "\n",
    "# Create the model data from the decoded contents\n",
    "fname = structs[list(structs.keys())[3]]\n",
    "modata = parser.create_data(fname)\n",
    "fmodel = files_data_style(modata)\n",
    "with open(fmodel) as fm:\n",
    "    mdata = json.load(fm)\n",
    "\n",
    "datstyle = sparser.create_style(fname, 'cartoon', 'residue')\n",
    "\n",
    "fstyle = files_data_style(datstyle)\n",
    "\n",
    "with open(fstyle) as sf:\n",
    "    data_style = json.load(sf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'name': 'N', 'chain': 'A', 'positions': [30.514, 14.219, 2.014], 'residue_index': 1, 'element': 'N', 'residue_name': 'THR248', 'serial': 0}\n",
      "{'name': 'CA', 'chain': 'A', 'positions': [29.356, 13.316, 1.944], 'residue_index': 1, 'element': 'C', 'residue_name': 'THR248', 'serial': 1}\n",
      "{'name': 'C', 'chain': 'A', 'positions': [28.082, 14.164, 1.848], 'residue_index': 1, 'element': 'C', 'residue_name': 'THR248', 'serial': 2}\n",
      "{'name': 'O', 'chain': 'A', 'positions': [28.175, 15.387, 1.95], 'residue_index': 1, 'element': 'O', 'residue_name': 'THR248', 'serial': 3}\n",
      "{'name': 'CB', 'chain': 'A', 'positions': [29.354, 12.378, 3.163], 'residue_index': 1, 'element': 'C', 'residue_name': 'THR248', 'serial': 4}\n",
      "{'name': 'OG1', 'chain': 'A', 'positions': [29.706, 13.115, 4.319], 'residue_index': 1, 'element': 'O', 'residue_name': 'THR248', 'serial': 5}\n",
      "{'name': 'CG2', 'chain': 'A', 'positions': [30.366, 11.237, 2.997], 'residue_index': 1, 'element': 'C', 'residue_name': 'THR248', 'serial': 6}\n",
      "{'name': 'H', 'chain': 'A', 'positions': [30.679, 14.626, 2.913], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 7}\n",
      "{'name': 'HA', 'chain': 'A', 'positions': [29.431, 12.707, 1.049], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 8}\n",
      "{'name': 'HB', 'chain': 'A', 'positions': [28.359, 11.944, 3.288], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 9}\n",
      "{'name': 'HG1', 'chain': 'A', 'positions': [29.077, 12.884, 5.023], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 10}\n",
      "{'name': 'HG21', 'chain': 'A', 'positions': [30.3, 10.565, 3.854], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 11}\n",
      "{'name': 'HG22', 'chain': 'A', 'positions': [31.384, 11.623, 2.933], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 12}\n",
      "{'name': 'HG23', 'chain': 'A', 'positions': [30.14, 10.667, 2.095], 'residue_index': 1, 'element': 'H', 'residue_name': 'THR248', 'serial': 13}\n",
      "{'name': 'N', 'chain': 'A', 'positions': [26.92, 13.552, 1.613], 'residue_index': 2, 'element': 'N', 'residue_name': 'LYS249', 'serial': 14}\n",
      "{'name': 'CA', 'chain': 'A', 'positions': [25.628, 14.231, 1.52], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 15}\n",
      "{'name': 'C', 'chain': 'A', 'positions': [24.647, 13.475, 2.408], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 16}\n",
      "{'name': 'O', 'chain': 'A', 'positions': [24.773, 12.258, 2.579], 'residue_index': 2, 'element': 'O', 'residue_name': 'LYS249', 'serial': 17}\n",
      "{'name': 'CB', 'chain': 'A', 'positions': [25.101, 14.268, 0.074], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 18}\n",
      "{'name': 'CG', 'chain': 'A', 'positions': [25.546, 15.468, -0.776], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 19}\n",
      "{'name': 'CD', 'chain': 'A', 'positions': [26.883, 15.303, -1.507], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 20}\n",
      "{'name': 'CE', 'chain': 'A', 'positions': [27.189, 16.519, -2.396], 'residue_index': 2, 'element': 'C', 'residue_name': 'LYS249', 'serial': 21}\n",
      "{'name': 'NZ', 'chain': 'A', 'positions': [26.308, 16.629, -3.58], 'residue_index': 2, 'element': 'N', 'residue_name': 'LYS249', 'serial': 22}\n",
      "{'name': 'H', 'chain': 'A', 'positions': [26.899, 12.539, 1.535], 'residue_index': 2, 'element': 'H', 'residue_name': 'LYS249', 'serial': 23}\n",
      "{'name': 'HA', 'chain': 'A', 'positions': [25.709, 15.254, 1.898], 'residue_index': 2, 'element': 'H', 'residue_name': 'LYS249', 'serial': 24}\n",
      "{'name': 'CE2', 'chain': 'B', 'positions': [-23.361, 2.242, 5.969], 'residue_index': 4719, 'element': 'C', 'residue_name': 'PHE364', 'serial': 74615}\n",
      "{'name': 'CZ', 'chain': 'B', 'positions': [-22.586, 2.989, 5.067], 'residue_index': 4719, 'element': 'C', 'residue_name': 'PHE364', 'serial': 74616}\n",
      "{'name': 'H', 'chain': 'B', 'positions': [-22.7, 6.961, 9.587], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74617}\n",
      "{'name': 'HA', 'chain': 'B', 'positions': [-25.102, 6.168, 8.051], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74618}\n",
      "{'name': 'HB2', 'chain': 'B', 'positions': [-22.824, 4.756, 9.478], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74619}\n",
      "{'name': 'HB3', 'chain': 'B', 'positions': [-24.357, 3.936, 9.359], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74620}\n",
      "{'name': 'HD1', 'chain': 'B', 'positions': [-22.365, 5.905, 6.823], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74621}\n",
      "{'name': 'HD2', 'chain': 'B', 'positions': [-24.318, 2.217, 7.892], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74622}\n",
      "{'name': 'HE1', 'chain': 'B', 'positions': [-21.626, 4.888, 4.706], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74623}\n",
      "{'name': 'HE2', 'chain': 'B', 'positions': [-23.633, 1.222, 5.735], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74624}\n",
      "{'name': 'HZ', 'chain': 'B', 'positions': [-22.256, 2.542, 4.139], 'residue_index': 4719, 'element': 'H', 'residue_name': 'PHE364', 'serial': 74625}\n",
      "{'name': 'N', 'chain': 'B', 'positions': [-26.597, 5.324, 10.124], 'residue_index': 4720, 'element': 'N', 'residue_name': 'PRO365', 'serial': 74626}\n",
      "{'name': 'CA', 'chain': 'B', 'positions': [-27.502, 5.229, 11.258], 'residue_index': 4720, 'element': 'C', 'residue_name': 'PRO365', 'serial': 74627}\n",
      "{'name': 'C', 'chain': 'B', 'positions': [-26.874, 4.329, 12.318], 'residue_index': 4720, 'element': 'C', 'residue_name': 'PRO365', 'serial': 74628}\n",
      "{'name': 'O', 'chain': 'B', 'positions': [-26.648, 4.821, 13.442], 'residue_index': 4720, 'element': 'O', 'residue_name': 'PRO365', 'serial': 74629}\n",
      "{'name': 'CB', 'chain': 'B', 'positions': [-28.814, 4.667, 10.709], 'residue_index': 4720, 'element': 'C', 'residue_name': 'PRO365', 'serial': 74630}\n",
      "{'name': 'CG', 'chain': 'B', 'positions': [-28.378, 3.874, 9.478], 'residue_index': 4720, 'element': 'C', 'residue_name': 'PRO365', 'serial': 74631}\n",
      "{'name': 'CD', 'chain': 'B', 'positions': [-27.151, 4.633, 8.968], 'residue_index': 4720, 'element': 'C', 'residue_name': 'PRO365', 'serial': 74632}\n",
      "{'name': 'HA', 'chain': 'B', 'positions': [-27.676, 6.219, 11.682], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74633}\n",
      "{'name': 'HB2', 'chain': 'B', 'positions': [-29.322, 4.037, 11.443], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74634}\n",
      "{'name': 'HB3', 'chain': 'B', 'positions': [-29.456, 5.494, 10.403], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74635}\n",
      "{'name': 'HG2', 'chain': 'B', 'positions': [-28.085, 2.868, 9.777], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74636}\n",
      "{'name': 'HG3', 'chain': 'B', 'positions': [-29.167, 3.832, 8.726], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74637}\n",
      "{'name': 'HD2', 'chain': 'B', 'positions': [-26.437, 3.937, 8.529], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74638}\n",
      "{'name': 'HD3', 'chain': 'B', 'positions': [-27.46, 5.371, 8.228], 'residue_index': 4720, 'element': 'H', 'residue_name': 'PRO365', 'serial': 74639}\n",
      "{'color': '#FA9600', 'visualization_type': 'cartoon'}\n",
      "74639\n",
      "74639\n"
     ]
    }
   ],
   "source": [
    "for atom_id in mdata['atoms'][0:25]:\n",
    "    print(atom_id)\n",
    "for atom_id in mdata['atoms'][-25:]:\n",
    "    print(atom_id)\n",
    "print( data_style['1'] )\n",
    "maxnum = 0\n",
    "for serial in data_style:\n",
    "    if int(serial)>maxnum :\n",
    "        maxnum = int(serial)\n",
    "maxserial = 0\n",
    "for atom_id in mdata['atoms']:\n",
    "    if int(atom_id['serial'])>maxserial :\n",
    "        maxserial = int(atom_id['serial'])\n",
    "        \n",
    "print(maxnum)\n",
    "print(maxserial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}, {'color': '#00CCCC', 'visualization_type': 'cartoon'}]\n"
     ]
    }
   ],
   "source": [
    "def set_color_cluster(structfile, cluster , mdata ):\n",
    "    #blue and yellow high contrast\n",
    "    #find match in annotation and blast ouput\n",
    "    #geno positions from cluster file\n",
    "    positions = [ int(p) for p in clusterdf.loc[cluster] if ~np.isnan(p) ]\n",
    "    #select struct\n",
    "    geno_annot = annotation[ structfile == annotation.struct ]\n",
    "    residues = {}\n",
    "    for i,r in geno_annot.iterrows():\n",
    "        if r.chain not in residues:\n",
    "            residues[r.chain] = []\n",
    "        for p in positions:\n",
    "            if p < r.qend and p > r.qstart:\n",
    "                print(p)\n",
    "                #match codon to amino acid position\n",
    "                #todo: triple check this!!!!\n",
    "                residues[r.chain].append( int( p - r.qstart / 3 ) )\n",
    "    \n",
    "    #map from geno pos to residue index\n",
    "    colors =[]\n",
    "    #intersect both\n",
    "    #list comp to generate residue positions\n",
    "    for atom_data in mdata['atoms']:\n",
    "        for chain in residues:\n",
    "            if atom_data['chain'] == chain  and atom_data['residue_index'] in residues[chain]:\n",
    "                #color select\n",
    "                #set the sidechains to stick view\n",
    "                colors.append({'color': '#CC0000', 'visualization_type': 'stick'})\n",
    "            else:\n",
    "                #color bg \n",
    "                colors.append({'color': '#00CCCC', 'visualization_type': 'cartoon'})\n",
    "    return colors\n",
    "\n",
    "colors = set_color_cluster( list(structs.keys())[3], 20 , mdata)\n",
    "print(colors[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']\n",
      "{'color': '#FA9600', 'visualization_type': 'cartoon'}\n"
     ]
    }
   ],
   "source": [
    "print(list(data_style.keys())[0:10])\n",
    "print(data_style['0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "external_stylesheets = ['https://codepen.io/chriddyp/pen/bWLwgP.css']\n",
    "\n",
    "app = dash.Dash(__name__, external_stylesheets=external_stylesheets)\n",
    "\n",
    "data_info = { os.path.abspath( structs[s])  : {\n",
    "        'name': 'covid prot'+str(i),\n",
    "        'description': 'test description',\n",
    "        'link':'test link' } for i,s in enumerate(structs) }\n",
    "stacktrack = {\n",
    "                    'type': 'STACK',\n",
    "                    'data': circos_graph_data['stack'],\n",
    "                    'config': {\n",
    "                        'innerRadius': .9 ,\n",
    "                        'outerRadius': 1.2,\n",
    "                        'thickness': 4,\n",
    "                        'margin': 0,\n",
    "                        'direction': 'out',\n",
    "                        'strokeWidth': 0,\n",
    "                        'opacity': 0.5,\n",
    "                        'tooltipContent': {'name': 'chr'}\n",
    "                        ,'color': {\n",
    "                            'conditional': {\n",
    "                                'end': 'end',\n",
    "                                'start': 'start',\n",
    "                                'value': [ n for n in range( 30000 , 1000)  ],\n",
    "                                'color': [\n",
    "                                    'red',\n",
    "                                    'black',\n",
    "                                    '#fff',\n",
    "                                    '#999',\n",
    "                                    '#BBB',\n",
    "                                ],\n",
    "                            }\n",
    "                            \n",
    "                        }\n",
    "                }\n",
    "}\n",
    "\n",
    "\n",
    "structtrack = {\n",
    "                    'type': 'STACK',\n",
    "                    'data': circos_graph_data['struct_stack'],\n",
    "                    'config': {\n",
    "                        'innerRadius': 0.7,\n",
    "                        'outerRadius': 0.9,\n",
    "                        'thickness': 4,\n",
    "                        'margin': 0,\n",
    "                        'direction': 'out',\n",
    "                        'strokeWidth': 0,\n",
    "                        'opacity': 0.5,\n",
    "                        'tooltipContent': {'name': 'chr'}\n",
    "                        }\n",
    "                }\n",
    "def header_colors():\n",
    "    return {\n",
    "        'bg_color': '#e7625f',\n",
    "        'font_color': 'white'\n",
    "    }\n",
    "\n",
    "def description():\n",
    "    return 'Molecule visualization in 3D - perfect for viewing ' \\\n",
    "           'biomolecules such as proteins, DNA and RNA. Includes ' \\\n",
    "           'stick, cartoon, and sphere representations.'\n",
    "\n",
    "app.layout = html.Div(\n",
    "        id='mol3d-body',\n",
    "        className = 'row',\n",
    "        children=[\n",
    "            \n",
    "######################### circos control pannel #####################################\n",
    "        html.Div([ \n",
    "       \n",
    "        html.Div([ \n",
    "        dashbio.Circos(\n",
    "        id='my-dashbio-circos',\n",
    "        layout= circos_graph_data['layout'],\n",
    "        selectEvent={\"0\": \"hover\", \"1\": \"click\", \"2\": \"both\"},\n",
    "        tracks=[\n",
    "           \n",
    "            {\n",
    "            'type': 'CHORDS',\n",
    "            'data': circos_graph_data[0],\n",
    "                \n",
    "            'config': {\n",
    "                'tooltipContent': {\n",
    "                    'source': 'source',\n",
    "                    'sourceID': 'id',\n",
    "                    'target': 'target',\n",
    "                    'targetID': 'id',\n",
    "                    'targetEnd': 'end'\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "            ,\n",
    "            stacktrack     \n",
    "            ,\n",
    "            structtrack\n",
    "            \n",
    "    \n",
    "        ] ) \n",
    "\n",
    "        ], id = 'circos' , className=\"six columns\"  )  \n",
    "            \n",
    "       \n",
    "        \n",
    "        ,\n",
    "\n",
    "        dcc.Store(\n",
    "                id='mol3d-color-storage',\n",
    "                data={}\n",
    "            ),\n",
    "    \n",
    "        html.Div(\n",
    "        id='mol3d-biomolecule-viewer',\n",
    "        children= [\n",
    "            \n",
    "############################# 3d viewer ####################################################\n",
    "        dash_bio.Molecule3dViewer(\n",
    "        id='mol-3d',\n",
    "        selectionType='atom',\n",
    "        modelData=mdata,\n",
    "        styles = data_style ,\n",
    "        selectedAtomIds=[],\n",
    "        backgroundOpacity='0',\n",
    "        atomLabelsShown=False,\n",
    "        )\n",
    "\n",
    "        \n",
    "        ] , className=\"six columns\")\n",
    "        ,\n",
    "\n",
    "####################### control pannel ############################################\n",
    "                    \n",
    "        html.Div( className=\"six columns\" , children = [\n",
    "        \n",
    "            html.Div(\n",
    "                id='mol3d-control-tabs',\n",
    "                className='control-tabs',\n",
    "                children=[\n",
    "                    dcc.Tabs(id='mol3d-tabs', value='what-is', children=[\n",
    "                        dcc.Tab(\n",
    "                            label='cluster select',\n",
    "                            value='what-is',\n",
    "                            children=html.Div(className='control-tab', children=[    \n",
    "                            \"Graph type:\",\n",
    "                            dcc.Dropdown(\n",
    "                                id='histogram-chords',\n",
    "                                options=[\n",
    "                                    {'label': x, 'value': x}\n",
    "                                    for x in list(circos_graph_data.keys()) if type(x) is int\n",
    "                                    ],\n",
    "                                value=0\n",
    "                                        ),\n",
    "                                \n",
    "                            \"Event data:\",    \n",
    "                            html.Div(id='circos-output')\n",
    "                                \n",
    "                            ] )\n",
    "                        )\n",
    "                        \n",
    "                        ,\n",
    "                        \n",
    "                        dcc.Tab(\n",
    "                            label='Structure select',\n",
    "                            value='upload-download',\n",
    "                            children=html.Div(className=' six columns', children=[\n",
    "                                html.Div(\n",
    "                                    title='Select molecule to view',\n",
    "                                    className=\"app-controls-block\",\n",
    "                                    children=[\n",
    "                                        html.Div(className='app-controls-name',\n",
    "                                                 children='Select structure'),\n",
    "                                        dcc.Dropdown(\n",
    "                                            id='dropdown-demostr',\n",
    "                                            options=[ { 'label':s,  'value':os.path.abspath(structs[s]) } for s in structs ],\n",
    "                                            value= os.path.abspath(structs[list(structs.keys())[3]])\n",
    "                                        ),\n",
    "                                    ],\n",
    "                                ),\n",
    "                            html.Div(id='mol3d-data-info')\n",
    "                            ])\n",
    "                        ),\n",
    "\n",
    "                        dcc.Tab(\n",
    "                            label='Structure Info and coloring',\n",
    "                            value='view-options',\n",
    "                            children=html.Div(className='control-tab', children=[\n",
    "                                # Textarea container to display the selected atoms\n",
    "                                html.Div(\n",
    "                                    title='view information about selected atoms '\n",
    "                                    'of biomolecule',\n",
    "                                    className=\"app-controls-block\",\n",
    "                                    id=\"mol3d-selection-display\",\n",
    "                                    children=[\n",
    "                                        html.P(\n",
    "                                            \"Selection\",\n",
    "                                            style={\n",
    "                                                'font-weight': 'bold',\n",
    "                                                'margin-bottom': '10px'\n",
    "                                            }\n",
    "                                        ),\n",
    "                                        html.Div(id='mol3d-selection-output'),\n",
    "                                    ]\n",
    "                                ),\n",
    "                                # Dropdown to select chain representation\n",
    "                                # (sticks, cartoon, sphere)\n",
    "                                html.Div(\n",
    "                                    title='select style for molecule representation',\n",
    "                                    className=\"app-controls-block\",\n",
    "                                    id='mol3d-style',\n",
    "                                    children=[\n",
    "                                        html.P(\n",
    "                                            'Style',\n",
    "                                            style={\n",
    "                                                'font-weight': 'bold',\n",
    "                                                'margin-bottom': '10px'\n",
    "                                            }\n",
    "                                        ),\n",
    "                                        dcc.Dropdown(\n",
    "                                            id='dropdown-styles',\n",
    "                                            options=[\n",
    "                                                {'label': 'Sticks', 'value': 'stick'},\n",
    "                                                {'label': 'Cartoon', 'value': 'cartoon'},\n",
    "                                                {'label': 'Spheres', 'value': 'sphere'},\n",
    "                                            ],\n",
    "                                            value='cartoon'\n",
    "                                        ),\n",
    "                                    ],\n",
    "                                ),\n",
    "\n",
    "                                # Dropdown to select color of representation\n",
    "                                html.Div(\n",
    "                                    title='select color scheme for viewing biomolecule',\n",
    "                                    className=\"app-controls-block\",\n",
    "                                    id='mol3d-style-color',\n",
    "                                    children=[\n",
    "                                        html.P(\n",
    "                                            'Color',\n",
    "                                            style={\n",
    "                                                'font-weight': 'bold',\n",
    "                                                'margin-bottom': '10px'\n",
    "                                            }\n",
    "                                        ),\n",
    "                                        dcc.Dropdown(\n",
    "                                            id='dropdown-style-color',\n",
    "                                            options=[\n",
    "                                                {'label': 'Atom',\n",
    "                                                 'value': 'atom'},\n",
    "                                                {'label': 'Residue identity',\n",
    "                                                 'value': 'residue'},\n",
    "                                                {'label': 'Residue type',\n",
    "                                                 'value': 'residue_type'},\n",
    "                                                {'label': 'Chain',\n",
    "                                                 'value': 'chain'},\n",
    "                                            ],\n",
    "                                            value='residue'\n",
    "                                        ),\n",
    "                                        dcc.Dropdown(\n",
    "                                            id='mol3d-coloring-key',\n",
    "                                            options=[]\n",
    "                                        ),\n",
    "\n",
    "                                    ],\n",
    "                                ),\n",
    "                                html.Div(\n",
    "                                    title='Customize molecule coloring.',\n",
    "                                    className=\"app-controls-block\",\n",
    "                                    children=[\n",
    "                                        html.P(\n",
    "                                            id='mol3d-customize-coloring',\n",
    "                                            style={\n",
    "                                                'font-weight': 'bold',\n",
    "                                                'margin-bottom': '10px'\n",
    "                                            }\n",
    "                                        ),\n",
    "                                        daq.ColorPicker(\n",
    "                                            id='mol3d-coloring-value',\n",
    "                                            size=315\n",
    "                                        ),\n",
    "                                    ]\n",
    "                                ),\n",
    "                            ]),\n",
    "                        ),\n",
    "\n",
    "                    ]),\n",
    "                ]  ),\n",
    "        ])\n",
    "        ],className=\"row\") , dcc.Store(id='structswitch' )   ])\n",
    "        \n",
    "\n",
    "\n",
    "########################circos functions ########################\n",
    "@app.callback(\n",
    "    \n",
    "    Output('circos-output', 'children'),\n",
    "    [Input('my-dashbio-circos', 'eventDatum')]\n",
    "    \n",
    ")\n",
    "def update_output(value):\n",
    "    if value is not None:\n",
    "        return [html.Div('{}: {}'.format(v.title(), value[v]))\n",
    "                for v in value.keys()]\n",
    "    return 'There are no event data. Click or hover on a data point to get more information.'\n",
    "\n",
    "####change the circos content\n",
    "@app.callback(\n",
    "    Output('my-dashbio-circos', 'tracks'),\n",
    "    [Input('histogram-chords', 'value')],\n",
    "    state=[State('my-dashbio-circos', 'tracks')]\n",
    ")\n",
    "def change_graph_type(value, current):   \n",
    "    current[0].update(\n",
    "            {\n",
    "                \n",
    "            'type': 'CHORDS',\n",
    "            'data': circos_graph_data[int(value)] ,\n",
    "            'config': {\n",
    "                'tooltipContent': {\n",
    "                    'source': 'source',\n",
    "                    'sourceID': 'id',\n",
    "                    'target': 'target',\n",
    "                    'targetID': 'id',\n",
    "                    'targetEnd': 'end'\n",
    "                }\n",
    "            }\n",
    "        }   \n",
    "    )\n",
    "    current[1].update(stacktrack)\n",
    "    current[2].update(structtrack)\n",
    "    return current\n",
    "            \n",
    "\n",
    "######### Function to create the modelData and style files for molecule visualization\n",
    "@app.callback(\n",
    "    Output('mol3d-data-info', 'children'),\n",
    "    [Input('dropdown-demostr', 'value')]\n",
    ")\n",
    "def show_data(molecule_selected):\n",
    "    if molecule_selected in data_info.keys():\n",
    "        mol = data_info[molecule_selected]\n",
    "        return [\n",
    "            html.H4(mol['name']),\n",
    "            mol['description'],\n",
    "            html.A(\n",
    "                '(source)',\n",
    "                href=mol['link']\n",
    "            )\n",
    "        ]\n",
    "    return ''\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Callback for updating dropdown options\n",
    "@app.callback(\n",
    "    Output('mol3d-coloring-key', 'options'),\n",
    "    [Input('dropdown-style-color', 'value')]\n",
    ")\n",
    "def update_color_options(mol_style):\n",
    "    color_dict_keys = {\n",
    "        'atom': list(sparser.ATOM_COLOR_DICT.keys()),\n",
    "        'residue': list(sparser.RESIDUE_COLOR_DICT.keys()),\n",
    "        'residue_type': list(sparser.RESIDUE_TYPE_COLOR_DICT.keys()),\n",
    "        'chain': list(sparser.CHAIN_COLOR_DICT.keys())\n",
    "    }\n",
    "    options = [{'label': k.upper(), 'value': k}\n",
    "               for k in color_dict_keys[mol_style]]\n",
    "    return options\n",
    "\n",
    "@app.callback(\n",
    "    [Output('structswitch','children'),Output(\"structswitch\", \"data\")],  \n",
    "    [ Input('dropdown-demostr', 'value'),Input('histogram-chords', 'value') , Input('mol-3d' , 'modelData') ] ,\n",
    ")\n",
    "def set_color_cluster( structfile, cluster , mdata ):\n",
    "    #blue and yellow high contrast\n",
    "    #find match in annotation and blast ouput\n",
    "    #geno positions from cluster file\n",
    "    positions = [ int(p) for p in clusterdf.loc[cluster] if ~np.isnan(p) ]\n",
    "    #select struct\n",
    "    print(structfile)\n",
    "    struct = structfile.split('/')[-1].replace('.pdb','').lower()\n",
    "    geno_annot = annotation[ struct == annotation.struct ]\n",
    "    print(geno_annot)\n",
    "    residues = {}\n",
    "    for i,r in geno_annot.iterrows():\n",
    "        if r.chain not in residues:\n",
    "            residues[r.chain] = []\n",
    "        for p in positions:\n",
    "            if p < r.qend and p > r.qstart:\n",
    "                print(p)\n",
    "                #match codon to amino acid position\n",
    "                #todo: triple check this!!!!\n",
    "                residues[r.chain].append( int( (p - r.qstart) / 3 ) )\n",
    "    \n",
    "    print(residues)\n",
    "    \n",
    "    currentselection = []\n",
    "    chaindiffs = {}\n",
    "    chaincount = 0\n",
    "    \n",
    "    print('chain offset' , chaindiffs)\n",
    "    \n",
    "    for atom_data in mdata['atoms']:\n",
    "        for chain in residues:\n",
    "            ###todo : triple check this\n",
    "            if atom_data['chain'] == chain  and int(atom_data['residue_name'][3:]) in residues[chain]:\n",
    "                currentselection.append(atom_data['serial'])\n",
    "            #color select\n",
    "    print('selection', currentselection)\n",
    "    return structfile , currentselection\n",
    "\n",
    "# Callback for molecule visualization based on uploaded PDB file\n",
    "@app.callback(\n",
    "    Output('mol3d-biomolecule-viewer', 'children'),\n",
    "    [Input(\"structswitch\", \"data\") , Input('dropdown-demostr', 'value'),\n",
    "    Input('dropdown-styles', 'value'),\n",
    "    Input('dropdown-style-color', 'value'),\n",
    "    Input('mol3d-color-storage', 'modified_timestamp')],\n",
    "    [State('mol3d-color-storage', 'data') ],\n",
    ")\n",
    "def use_upload( selection , demostr, mol_style, color_style, mt, custom_colors  ):\n",
    "    print(selection)\n",
    "    fname = demostr\n",
    "    \n",
    "    copy2(demostr, './str.pdb')\n",
    "    fname = './str.pdb'\n",
    "    # Create the model data from the decoded contents\n",
    "    \n",
    "    print(custom_colors)\n",
    "    modata = parser.create_data(fname)\n",
    "    fmodel = files_data_style(modata)\n",
    "    with open(fmodel) as fm:\n",
    "        mdata = json.load(fm)\n",
    "    #all blue\n",
    "    datstyle = sparser.create_style(fname, mol_style, color_style, **custom_colors)\n",
    "    fstyle = files_data_style(datstyle)\n",
    "    with open(fstyle) as sf:\n",
    "        data_style = json.load(sf)\n",
    "    \n",
    "    print('selection finish' , selection)\n",
    "    data_style = [ {'color': 'blue', 'visualization_type': 'cartoon'}  if i not in selection else  {'color': 'yellow', 'visualization_type': 'cartoon'}  for i,atom in enumerate(data_style) ]\n",
    "    \n",
    "    # Return the new molecule visualization container\n",
    "    return dash_bio.Molecule3dViewer(\n",
    "        id='mol-3d',\n",
    "        selectionType='atom',\n",
    "        modelData=mdata,\n",
    "        styles=data_style,\n",
    "        selectedAtomIds= [] ,\n",
    "        backgroundOpacity='0',\n",
    "        atomLabelsShown=False,\n",
    "        )\n",
    "\n",
    "#### add struct select callback from circos#######\n",
    "# Callback to print details of each selected atom of the biomolecule\n",
    "@app.callback(\n",
    "    Output(\"mol3d-selection-output\", \"children\"),\n",
    "    [Input(\"mol-3d\", \"selectedAtomIds\"),\n",
    "     Input(\"mol-3d\", \"modelData\")]\n",
    ")\n",
    "def selout(selected_atom_ids, model_data):\n",
    "    residue_summary = []\n",
    "    for atom_id in selected_atom_ids:\n",
    "        res_info = model_data['atoms'][atom_id]\n",
    "        residues = {\n",
    "            \"residue\": res_info['residue_name'],\n",
    "            \"atom\": res_info['name'],\n",
    "            \"chain\": res_info['chain'],\n",
    "            \"xyz\": res_info['positions']}\n",
    "        \n",
    "        residue_summary += [html.P('{}: {}'.format(\n",
    "            key, str(residues[key]))) for key in residues]\n",
    "        residue_summary.append(html.Br())\n",
    "    if len(residue_summary) == 0:\n",
    "        residue_summary.append(\"No atoms have been selected. Click \\\n",
    "        on an atom to select it.\")\n",
    "    return html.Div( residue_summary )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8050/\n",
      "\n",
      " Warning: This is a development server. Do not use app.run_server\n",
      " Warning: This is a development server. Do not use app.run_server\n",
      " Warning: This is a development server. Do not use app.run_server\n",
      " Warning: This is a development server. Do not use app.run_server\n",
      " in production, use a production WSGI server like gunicorn instead.\n",
      "\n",
      " in production, use a production WSGI server like gunicorn instead.\n",
      "\n",
      " in production, use a production WSGI server like gunicorn instead.\n",
      "\n",
      " in production, use a production WSGI server like gunicorn instead.\n",
      "\n",
      " * Serving Flask app \"__main__\" (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)\n"
     ]
    }
   ],
   "source": [
    "app.run_server( port=8050 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
